File: mac/Core/KM/KmPayload.swift
import Foundation

public enum KmPayload {
    public struct MouseMove: Equatable {
        public let x: Int32
        public let y: Int32

        public init(x: Int32, y: Int32) {
            self.x = x
            self.y = y
        }
    }

    public struct MouseButton: Equatable {
        public let button: UInt8
        public let isDown: Bool

        public init(button: UInt8, isDown: Bool) {
            self.button = button
            self.isDown = isDown
        }
    }

    public struct MouseScroll: Equatable {
        public let deltaX: Int32
        public let deltaY: Int32

        public init(deltaX: Int32, deltaY: Int32) {
            self.deltaX = deltaX
            self.deltaY = deltaY
        }
    }

    public struct Key: Equatable {
        public let keyCode: UInt16
        public let flags: UInt64

        public init(keyCode: UInt16, flags: UInt64) {
            self.keyCode = keyCode
            self.flags = flags
        }
    }

    public static func encodeMouseMove(_ move: MouseMove) -> Data {
        var data = Data(capacity: 8)
        data.appendLE(move.x)
        data.appendLE(move.y)
        return data
    }

    public static func decodeMouseMove(_ data: Data) throws -> MouseMove {
        guard data.count >= 8 else { throw BrpCodecError.truncated }
        return MouseMove(x: data.readLE(at: 0), y: data.readLE(at: 4))
    }

    public static func encodeMouseButton(_ button: MouseButton) -> Data {
        return Data([button.button, button.isDown ? 1 : 0])
    }

    public static func decodeMouseButton(_ data: Data) throws -> MouseButton {
        guard data.count >= 2 else { throw BrpCodecError.truncated }
        return MouseButton(button: data.readU8(at: 0), isDown: data.readU8(at: 1) != 0)
    }

    public static func encodeMouseScroll(_ scroll: MouseScroll) -> Data {
        var data = Data(capacity: 8)
        data.appendLE(scroll.deltaX)
        data.appendLE(scroll.deltaY)
        return data
    }

    public static func decodeMouseScroll(_ data: Data) throws -> MouseScroll {
        guard data.count >= 8 else { throw BrpCodecError.truncated }
        return MouseScroll(deltaX: data.readLE(at: 0), deltaY: data.readLE(at: 4))
    }

    public static func encodeKey(_ key: Key) -> Data {
        var data = Data(capacity: 10)
        data.appendLE(key.keyCode)
        data.appendLE(key.flags)
        return data
    }

    public static func decodeKey(_ data: Data) throws -> Key {
        guard data.count >= 10 else { throw BrpCodecError.truncated }
        return Key(keyCode: data.readLE(at: 0), flags: data.readLE(at: 2))
    }

    public static func encodeModifiers(flags: UInt64) -> Data {
        var data = Data(capacity: 8)
        data.appendLE(flags)
        return data
    }

    public static func decodeModifiers(_ data: Data) throws -> UInt64 {
        guard data.count >= 8 else { throw BrpCodecError.truncated }
        return data.readLE(at: 0)
    }
}

File: windows/Core/Km/KmPayload.cs
using System.Buffers.Binary;
using BarelyReal.Core.Protocol;

namespace BarelyReal.Core.Km;

public static class KmPayload
{
    public readonly record struct MouseMove(int X, int Y);
    public readonly record struct MouseButton(byte Button, bool IsDown);
    public readonly record struct MouseScroll(int DeltaX, int DeltaY);
    public readonly record struct Key(ushort KeyCode, ulong Flags);

    public static byte[] EncodeMouseMove(MouseMove move)
    {
        var payload = new byte[8];
        BinaryPrimitives.WriteInt32LittleEndian(payload.AsSpan(0, 4), move.X);
        BinaryPrimitives.WriteInt32LittleEndian(payload.AsSpan(4, 4), move.Y);
        return payload;
    }

    public static MouseMove DecodeMouseMove(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 8)
            throw new BrpCodecException(BrpCodecError.Truncated);

        return new MouseMove(
            BinaryPrimitives.ReadInt32LittleEndian(payload[..4]),
            BinaryPrimitives.ReadInt32LittleEndian(payload.Slice(4, 4)));
    }

    public static byte[] EncodeMouseButton(MouseButton button) =>
        new[] { button.Button, button.IsDown ? (byte)1 : (byte)0 };

    public static MouseButton DecodeMouseButton(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 2)
            throw new BrpCodecException(BrpCodecError.Truncated);

        return new MouseButton(payload[0], payload[1] != 0);
    }

    public static byte[] EncodeMouseScroll(MouseScroll scroll)
    {
        var payload = new byte[8];
        BinaryPrimitives.WriteInt32LittleEndian(payload.AsSpan(0, 4), scroll.DeltaX);
        BinaryPrimitives.WriteInt32LittleEndian(payload.AsSpan(4, 4), scroll.DeltaY);
        return payload;
    }

    public static MouseScroll DecodeMouseScroll(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 8)
            throw new BrpCodecException(BrpCodecError.Truncated);

        return new MouseScroll(
            BinaryPrimitives.ReadInt32LittleEndian(payload[..4]),
            BinaryPrimitives.ReadInt32LittleEndian(payload.Slice(4, 4)));
    }

    public static byte[] EncodeKey(Key key)
    {
        var payload = new byte[10];
        BinaryPrimitives.WriteUInt16LittleEndian(payload.AsSpan(0, 2), key.KeyCode);
        BinaryPrimitives.WriteUInt64LittleEndian(payload.AsSpan(2, 8), key.Flags);
        return payload;
    }

    public static Key DecodeKey(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 10)
            throw new BrpCodecException(BrpCodecError.Truncated);

        return new Key(
            BinaryPrimitives.ReadUInt16LittleEndian(payload[..2]),
            BinaryPrimitives.ReadUInt64LittleEndian(payload.Slice(2, 8)));
    }

    public static byte[] EncodeModifiers(ulong flags)
    {
        var payload = new byte[8];
        BinaryPrimitives.WriteUInt64LittleEndian(payload, flags);
        return payload;
    }

    public static ulong DecodeModifiers(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 8)
            throw new BrpCodecException(BrpCodecError.Truncated);

        return BinaryPrimitives.ReadUInt64LittleEndian(payload[..8]);
    }
}
